File: sdk/core/ios/Sources/Mediation/FallbackWaterfall.swift
import Foundation

/// FallbackWaterfall - Implements a waterfall mediation strategy for ad loading.
///
/// When one ad source fails, it automatically falls back to the next source
/// in the waterfall. This ensures maximum fill rate by trying multiple
/// demand partners in priority order.
public final class FallbackWaterfall<T> {
    
    /// Default timeout for source loading
    public let defaultTimeout: TimeInterval
    
    /// Creates a new FallbackWaterfall
    public init(defaultTimeout: TimeInterval = 5.0) {
        self.defaultTimeout = defaultTimeout
    }
    
    /// Represents an ad source in the waterfall
    public struct WaterfallSource {
        public let id: String
        public let priority: Int
        public let timeout: TimeInterval
        public let loader: (@escaping (WaterfallResult<T>) -> Void) -> Void
        
        public init(
            id: String,
            priority: Int,
            timeout: TimeInterval,
            loader: @escaping (@escaping (WaterfallResult<T>) -> Void) -> Void
        ) {
            self.id = id
            self.priority = priority
            self.timeout = timeout
            self.loader = loader
        }
    }
    
    /// Result of attempting to load from a source
    public enum WaterfallResult<T> {
        case success(T)
        case noFill(reason: String)
        case error(Error)
        case timeout
    }
    
    /// Result type for attempt tracking
    public enum AttemptResultType: String {
        case success
        case noFill
        case error
        case timeout
        case skipped
    }
    
    /// Details about each attempt in the waterfall
    public struct AttemptDetail {
        public let sourceId: String
        public let priority: Int
        public let durationMs: Int64
        public let result: AttemptResultType
    }
    
    /// Waterfall execution result with metadata
    public struct ExecutionResult {
        public let result: WaterfallResult<T>
        public let sourceId: String
        public let attemptsCount: Int
        public let totalDurationMs: Int64
        public let attemptDetails: [AttemptDetail]
    }
    
    /// Performance statistics for a source
    public class SourceStats {
        private let lock = NSLock()
        
        private var _successCount: Int = 0
        private var _failureCount: Int = 0
        private var _timeoutCount: Int = 0
        private var _noFillCount: Int = 0
        private var _totalLatencyMs: Int64 = 0
        
        public var successCount: Int {
            lock.lock()
            defer { lock.unlock() }
            return _successCount
        }
        
        public var failureCount: Int {
            lock.lock()
            defer { lock.unlock() }
            return _failureCount
        }
        
        public var timeoutCount: Int {
            lock.lock()
            defer { lock.unlock() }
            return _timeoutCount
        }
        
        public var noFillCount: Int {
            lock.lock()
            defer { lock.unlock() }
            return _noFillCount
        }
        
        public var totalAttempts: Int {
            lock.lock()
            defer { lock.unlock() }
            return _successCount + _failureCount + _timeoutCount + _noFillCount
        }
        
        public var successRate: Float {
            let total = totalAttempts
            guard total > 0 else { return 0 }
            return Float(successCount) / Float(total)
        }
        
        public var averageLatencyMs: Int64 {
            let total = totalAttempts
            guard total > 0 else { return 0 }
            lock.lock()
            defer { lock.unlock() }
            return _totalLatencyMs / Int64(total)
        }
        
        func recordSuccess(latencyMs: Int64) {
            lock.lock()
            _successCount += 1
            _totalLatencyMs += latencyMs
            lock.unlock()
        }
        
        func recordFailure(latencyMs: Int64) {
            lock.lock()
            _failureCount += 1
            _totalLatencyMs += latencyMs
            lock.unlock()
        }
        
        func recordTimeout(latencyMs: Int64) {
            lock.lock()
            _timeoutCount += 1
            _totalLatencyMs += latencyMs
            lock.unlock()
        }
        
        func recordNoFill(latencyMs: Int64) {
            lock.lock()
            _noFillCount += 1
            _totalLatencyMs += latencyMs
            lock.unlock()
        }
    }
    
    private var sourceStats: [String: SourceStats] = [:]
    private let statsLock = NSLock()
    
    /// Executes the waterfall, trying each source in priority order
    public func execute(
        sources: [WaterfallSource],
        completion: @escaping (ExecutionResult) -> Void
    ) {
        let startTime = Date()
        var attemptDetails: [AttemptDetail] = []
        let sortedSources = sources.sorted { $0.priority < $1.priority }
        
        executeNext(
            sources: sortedSources,
            index: 0,
            startTime: startTime,
            attemptDetails: attemptDetails,
            completion: completion
        )
    }
    
    private func executeNext(
        sources: [WaterfallSource],
        index: Int,
        startTime: Date,
        attemptDetails: [AttemptDetail],
        completion: @escaping (ExecutionResult) -> Void
    ) {
        var attemptDetails = attemptDetails
        
        guard index < sources.count else {
            // All sources exhausted
            let totalDuration = Int64(Date().timeIntervalSince(startTime) * 1000)
            completion(ExecutionResult(
                result: .noFill(reason: "All \(sources.count) sources exhausted"),
                sourceId: sources.last?.id ?? "none",
                attemptsCount: attemptDetails.count,
                totalDurationMs: totalDuration,
                attemptDetails: attemptDetails
            ))
            return
        }
        
        let source = sources[index]
        let attemptStart = Date()
        var completed = false
        
        // Set up timeout
        let timeoutWorkItem = DispatchWorkItem { [weak self] in
            guard !completed else { return }
            completed = true
            
            let duration = Int64(Date().timeIntervalSince(attemptStart) * 1000)
            self?.recordStats(sourceId: source.id, result: .timeout, durationMs: duration)
            
            attemptDetails.append(AttemptDetail(
                sourceId: source.id,
                priority: source.priority,
                durationMs: duration,
                result: .timeout
            ))
            
            // Try next source
            self?.executeNext(
                sources: sources,
                index: index + 1,
                startTime: startTime,
                attemptDetails: attemptDetails,
                completion: completion
            )
        }
        
        DispatchQueue.global().asyncAfter(deadline: .now() + source.timeout, execute: timeoutWorkItem)
        
        // Execute loader
        source.loader { [weak self] result in
            guard !completed else { return }
            completed = true
            timeoutWorkItem.cancel()
            
            let duration = Int64(Date().timeIntervalSince(attemptStart) * 1000)
            self?.recordStats(sourceId: source.id, result: result, durationMs: duration)
            
            let attemptResultType: AttemptResultType
            switch result {
            case .success: attemptResultType = .success
            case .noFill: attemptResultType = .noFill
            case .error: attemptResultType = .error
            case .timeout: attemptResultType = .timeout
            }
            
            attemptDetails.append(AttemptDetail(
                sourceId: source.id,
                priority: source.priority,
                durationMs: duration,
                result: attemptResultType
            ))
            
            switch result {
            case .success:
                let totalDuration = Int64(Date().timeIntervalSince(startTime) * 1000)
                completion(ExecutionResult(
                    result: result,
                    sourceId: source.id,
                    attemptsCount: attemptDetails.count,
                    totalDurationMs: totalDuration,
                    attemptDetails: attemptDetails
                ))
                
            case .noFill, .error, .timeout:
                self?.executeNext(
                    sources: sources,
                    index: index + 1,
                    startTime: startTime,
                    attemptDetails: attemptDetails,
                    completion: completion
                )
            }
        }
    }
    
    /// Gets statistics for a specific source
    public func getStats(sourceId: String) -> SourceStats? {
        statsLock.lock()
        defer { statsLock.unlock() }
        return sourceStats[sourceId]
    }
    
    /// Gets statistics for all sources
    public func getAllStats() -> [String: SourceStats] {
        statsLock.lock()
        defer { statsLock.unlock() }
        return sourceStats
    }
    
    /// Clears all recorded statistics
    public func clearStats() {
        statsLock.lock()
        sourceStats.removeAll()
        statsLock.unlock()
    }
    
    /// Creates a source with the given parameters
    public func createSource(
        id: String,
        priority: Int,
        timeout: TimeInterval? = nil,
        loader: @escaping (@escaping (WaterfallResult<T>) -> Void) -> Void
    ) -> WaterfallSource {
        WaterfallSource(
            id: id,
            priority: priority,
            timeout: timeout ?? defaultTimeout,
            loader: loader
        )
    }
    
    private func recordStats(sourceId: String, result: WaterfallResult<T>, durationMs: Int64) {
        statsLock.lock()
        let stats = sourceStats[sourceId] ?? SourceStats()
        sourceStats[sourceId] = stats
        statsLock.unlock()
        
        switch result {
        case .success: stats.recordSuccess(latencyMs: durationMs)
        case .noFill: stats.recordNoFill(latencyMs: durationMs)
        case .error: stats.recordFailure(latencyMs: durationMs)
        case .timeout: stats.recordTimeout(latencyMs: durationMs)
        }
    }
}

/// Builder for creating waterfall configurations
public class WaterfallBuilder<T> {
    private let defaultTimeout: TimeInterval
    private var sources: [FallbackWaterfall<T>.WaterfallSource] = []
    
    public init(defaultTimeout: TimeInterval = 5.0) {
        self.defaultTimeout = defaultTimeout
    }
    
    @discardableResult
    public func addSource(
        id: String,
        priority: Int,
        timeout: TimeInterval? = nil,
        loader: @escaping (@escaping (FallbackWaterfall<T>.WaterfallResult<T>) -> Void) -> Void
    ) -> WaterfallBuilder<T> {
        sources.append(FallbackWaterfall<T>.WaterfallSource(
            id: id,
            priority: priority,
            timeout: timeout ?? defaultTimeout,
            loader: loader
        ))
        return self
    }
    
    public func build() -> [FallbackWaterfall<T>.WaterfallSource] {
        sources
    }
}

File: sdk/core/android/src/main/kotlin/mediation/FallbackWaterfall.kt
package com.rivalapex.sdk.mediation

import kotlinx.coroutines.*
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.ConcurrentHashMap

/**
 * FallbackWaterfall - Implements a waterfall mediation strategy for ad loading.
 *
 * When one ad source fails, it automatically falls back to the next source
 * in the waterfall. This ensures maximum fill rate by trying multiple
 * demand partners in priority order.
 *
 * Features:
 * - Priority-based waterfall ordering
 * - Configurable timeout per source
 * - Automatic failover on error, timeout, or no-fill
 * - Performance tracking per source
 * - Concurrent loading with ordered response handling
 */
class FallbackWaterfall<T>(
    private val defaultTimeout: Long = 5000L
) {
    
    /**
     * Represents an ad source in the waterfall
     */
    data class WaterfallSource<T>(
        val id: String,
        val priority: Int,
        val timeoutMs: Long,
        val loader: suspend () -> WaterfallResult<T>
    )
    
    /**
     * Result of attempting to load from a source
     */
    sealed class WaterfallResult<out T> {
        data class Success<T>(val data: T) : WaterfallResult<T>()
        data class NoFill(val reason: String = "No ad available") : WaterfallResult<Nothing>()
        data class Error(val error: Throwable) : WaterfallResult<Nothing>()
        object Timeout : WaterfallResult<Nothing>()
    }
    
    /**
     * Waterfall execution result with metadata
     */
    data class ExecutionResult<T>(
        val result: WaterfallResult<T>,
        val sourceId: String,
        val attemptsCount: Int,
        val totalDurationMs: Long,
        val attemptDetails: List<AttemptDetail>
    )
    
    /**
     * Details about each attempt in the waterfall
     */
    data class AttemptDetail(
        val sourceId: String,
        val priority: Int,
        val durationMs: Long,
        val result: AttemptResultType
    )
    
    enum class AttemptResultType {
        SUCCESS, NO_FILL, ERROR, TIMEOUT, SKIPPED
    }
    
    /**
     * Performance statistics for a source
     */
    data class SourceStats(
        val successCount: AtomicInteger = AtomicInteger(0),
        val failureCount: AtomicInteger = AtomicInteger(0),
        val timeoutCount: AtomicInteger = AtomicInteger(0),
        val noFillCount: AtomicInteger = AtomicInteger(0),
        val totalLatencyMs: AtomicInteger = AtomicInteger(0)
    ) {
        val totalAttempts: Int
            get() = successCount.get() + failureCount.get() + timeoutCount.get() + noFillCount.get()
        
        val successRate: Float
            get() = if (totalAttempts > 0) successCount.get().toFloat() / totalAttempts else 0f
        
        val averageLatencyMs: Long
            get() = if (totalAttempts > 0) totalLatencyMs.get().toLong() / totalAttempts else 0
    }
    
    private val sourceStats = ConcurrentHashMap<String, SourceStats>()
    
    /**
     * Executes the waterfall, trying each source in priority order until
     * one succeeds or all sources are exhausted.
     *
     * @param sources List of sources to try, ordered by priority
     * @return ExecutionResult containing the final result and metadata
     */
    suspend fun execute(sources: List<WaterfallSource<T>>): ExecutionResult<T> {
        val startTime = System.currentTimeMillis()
        val attemptDetails = mutableListOf<AttemptDetail>()
        val sortedSources = sources.sortedBy { it.priority }
        
        for (source in sortedSources) {
            val attemptStart = System.currentTimeMillis()
            
            val result = try {
                withTimeout(source.timeoutMs) {
                    source.loader()
                }
            } catch (e: TimeoutCancellationException) {
                WaterfallResult.Timeout
            } catch (e: CancellationException) {
                throw e // Rethrow cancellation
            } catch (e: Exception) {
                WaterfallResult.Error(e)
            }
            
            val attemptDuration = System.currentTimeMillis() - attemptStart
            recordStats(source.id, result, attemptDuration)
            
            val attemptResultType = when (result) {
                is WaterfallResult.Success -> AttemptResultType.SUCCESS
                is WaterfallResult.NoFill -> AttemptResultType.NO_FILL
                is WaterfallResult.Error -> AttemptResultType.ERROR
                is WaterfallResult.Timeout -> AttemptResultType.TIMEOUT
            }
            
            attemptDetails.add(
                AttemptDetail(
                    sourceId = source.id,
                    priority = source.priority,
                    durationMs = attemptDuration,
                    result = attemptResultType
                )
            )
            
            if (result is WaterfallResult.Success) {
                return ExecutionResult(
                    result = result,
                    sourceId = source.id,
                    attemptsCount = attemptDetails.size,
                    totalDurationMs = System.currentTimeMillis() - startTime,
                    attemptDetails = attemptDetails
                )
            }
        }
        
        // All sources exhausted
        val totalDuration = System.currentTimeMillis() - startTime
        return ExecutionResult(
            result = WaterfallResult.NoFill("All ${sortedSources.size} sources exhausted"),
            sourceId = sortedSources.lastOrNull()?.id ?: "none",
            attemptsCount = attemptDetails.size,
            totalDurationMs = totalDuration,
            attemptDetails = attemptDetails
        )
    }
    
    /**
     * Executes waterfall with optional parallel preloading of lower-priority sources.
     * This can reduce latency when a higher-priority source returns no-fill quickly.
     */
    suspend fun executeWithParallelPreload(
        sources: List<WaterfallSource<T>>,
        preloadCount: Int = 1
    ): ExecutionResult<T> = coroutineScope {
        val startTime = System.currentTimeMillis()
        val attemptDetails = mutableListOf<AttemptDetail>()
        val sortedSources = sources.sortedBy { it.priority }
        
        for ((index, source) in sortedSources.withIndex()) {
            // Start preloading next sources in background
            val preloadJobs = sortedSources
                .drop(index + 1)
                .take(preloadCount)
                .map { preloadSource ->
                    async(start = CoroutineStart.LAZY) {
                        try {
                            withTimeout(preloadSource.timeoutMs) {
                                preloadSource.loader()
                            }
                        } catch (e: Exception) {
                            WaterfallResult.Error(e)
                        }
                    }
                }
            
            // Execute current source
            val attemptStart = System.currentTimeMillis()
            
            val result = try {
                withTimeout(source.timeoutMs) {
                    source.loader()
                }
            } catch (e: TimeoutCancellationException) {
                WaterfallResult.Timeout
            } catch (e: CancellationException) {
                // Cancel preload jobs on cancellation
                preloadJobs.forEach { it.cancel() }
                throw e
            } catch (e: Exception) {
                WaterfallResult.Error(e)
            }
            
            val attemptDuration = System.currentTimeMillis() - attemptStart
            recordStats(source.id, result, attemptDuration)
            
            val attemptResultType = when (result) {
                is WaterfallResult.Success -> AttemptResultType.SUCCESS
                is WaterfallResult.NoFill -> AttemptResultType.NO_FILL
                is WaterfallResult.Error -> AttemptResultType.ERROR
                is WaterfallResult.Timeout -> AttemptResultType.TIMEOUT
            }
            
            attemptDetails.add(
                AttemptDetail(
                    sourceId = source.id,
                    priority = source.priority,
                    durationMs = attemptDuration,
                    result = attemptResultType
                )
            )
            
            if (result is WaterfallResult.Success) {
                preloadJobs.forEach { it.cancel() }
                return@coroutineScope ExecutionResult(
                    result = result,
                    sourceId = source.id,
                    attemptsCount = attemptDetails.size,
                    totalDurationMs = System.currentTimeMillis() - startTime,
                    attemptDetails = attemptDetails
                )
            }
        }
        
        // All sources exhausted
        val totalDuration = System.currentTimeMillis() - startTime
        ExecutionResult(
            result = WaterfallResult.NoFill("All ${sortedSources.size} sources exhausted"),
            sourceId = sortedSources.lastOrNull()?.id ?: "none",
            attemptsCount = attemptDetails.size,
            totalDurationMs = totalDuration,
            attemptDetails = attemptDetails
        )
    }
    
    /**
     * Gets statistics for a specific source.
     */
    fun getStats(sourceId: String): SourceStats? = sourceStats[sourceId]
    
    /**
     * Gets statistics for all sources.
     */
    fun getAllStats(): Map<String, SourceStats> = sourceStats.toMap()
    
    /**
     * Clears all recorded statistics.
     */
    fun clearStats() {
        sourceStats.clear()
    }
    
    /**
     * Creates a source with the given parameters.
     */
    fun createSource(
        id: String,
        priority: Int,
        timeoutMs: Long = defaultTimeout,
        loader: suspend () -> WaterfallResult<T>
    ) = WaterfallSource(id, priority, timeoutMs, loader)
    
    private fun recordStats(sourceId: String, result: WaterfallResult<T>, durationMs: Long) {
        val stats = sourceStats.getOrPut(sourceId) { SourceStats() }
        stats.totalLatencyMs.addAndGet(durationMs.toInt())
        
        when (result) {
            is WaterfallResult.Success -> stats.successCount.incrementAndGet()
            is WaterfallResult.NoFill -> stats.noFillCount.incrementAndGet()
            is WaterfallResult.Error -> stats.failureCount.incrementAndGet()
            is WaterfallResult.Timeout -> stats.timeoutCount.incrementAndGet()
        }
    }
}

/**
 * Builder for creating a waterfall configuration.
 */
class WaterfallBuilder<T>(private val defaultTimeout: Long = 5000L) {
    private val sources = mutableListOf<FallbackWaterfall.WaterfallSource<T>>()
    
    fun addSource(
        id: String,
        priority: Int,
        timeoutMs: Long = defaultTimeout,
        loader: suspend () -> FallbackWaterfall.WaterfallResult<T>
    ): WaterfallBuilder<T> {
        sources.add(FallbackWaterfall.WaterfallSource(id, priority, timeoutMs, loader))
        return this
    }
    
    fun build(): List<FallbackWaterfall.WaterfallSource<T>> = sources.toList()
}
